File: src/converter/plugins/convertLiteral.ts
import {ConverterPlugin, createSimplePlugin} from "../plugin.js";
import ts, {Node} from "typescript";
import {Render} from "../render.js";
import {convertPrimitive} from "./convertPrimitive.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";

export function convertLiteral<TNode extends Node = Node>(
    predicate: (node: Node) => node is TNode,
    render: Render<TNode>
): ConverterPlugin {
    const primitivePlugin = convertPrimitive(predicate, render)

    return createSimplePlugin((node, context, render) => {
        const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

        if (ts.isPrefixUnaryExpression(node) && predicate(node.operand)) {
            const result = primitivePlugin.render(node.operand, context, render)
            if (result === null) return null

            return `${result} /* ${typeScriptService?.printNode(node)} */`
        }

        const result = primitivePlugin.render(node, context, render)
        if (result === null) return null

        return `${result} /* ${typeScriptService?.printNode(node)} */`
    })
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertLiteral.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.Plugin
import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.Node
import typescript.isPrefixUnaryExpression

fun convertLiteral(
    predicate: (node: Node) -> Boolean,
    render: (node: Node) -> String?
): Plugin {
    val primitivePlugin = convertPrimitive(predicate, render)

    return createPlugin plugin@{ node, context, pluginRender ->
        val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

        if (isPrefixUnaryExpression(node) && predicate(node.operand)) {
            val result = primitivePlugin.render(node.operand, context, pluginRender)
            if (result === null) return@plugin null

            "$result /* ${typeScriptService?.printNode(node)} */"
        } else {
            val result = primitivePlugin.render(node, context, pluginRender)
            if (result === null) return@plugin null

            "$result /* ${typeScriptService?.printNode(node)} */"
        }
    }
}
